### PR TITLE
Low hanging fruit and the jump to R v4

### LINKED ISSUE #150: Remedy Travis config warnings
Deprecation warnings for build arguments that need to be addressed.

Screen Shot 2020-04-25 at 10 56 42 AM

### PR DESCRIPTION
**Pull request by nathancday**
This is PR in progress with remedies for small open issues for new functionality:

* `box_current_version()` from issue #142, hits the Version API similar to `box_previous_versions()` but just returns an integer >=1 
* Travis's warnings have been fixed #150 
* `box_saveRDS()/`box_readRDS()` from #110 is happening

### CODE REVIEW

**Comment by nathancday** [comment]
Adds support for #110

**Comment by nathancday** [comment]
Hey @ijlyttle this branch covers a lot of open issues and brings us up to speed for R 4.0.

I think it's time to do a CRAN release and I would like to submit it by the end of July.

Not all of this code has to go into the release, but most of the changes are on new functions, so there is low risk including them. Ideally I'd say we merge this and #155, then cut it.

What do you think?

File: R/boxr_file_versions.R [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
-#' versions on Box. No information about the current version of the file is
-#' returned. If the version of a file is one, then NULL is returned invisibly
+#' versions on Box, the column `file_version_id` can be passed to `version_id`
+#' in `box_dl()` and `box_read()`.
+#' If the version of a file is one, then NULL is returned invisibly
 #' along with a helpful message.
 #' 
-#' The returned `data.frame` contains a variable, `file_version_id`, 
-#' which you can use with [box_dl()].
+#' * `box_current_version()` returns a `integer`, starting from 1, which can be passed
```
Comment (ijlyttle) [inline R/boxr_file_versions.R:14]: how about `box_get_current_version()`? at some point also to rename (and soft-deprecate) `box_previous_versions()` to `box_file_get_info()` (I'd have to take a look at what gets returned)?

File: R/boxr__internal_verb_exit.R [hunk 2]
```diff
@@ -49,17 +49,38 @@ returnDwOp <- function(op_detail) {
...
   file_list <- 
     lapply(
       items_list,
       function(item)
         suppressWarnings(
-          data.frame(dplyr::bind_rows(lapply(
-            op_detail$files, function(x) data.frame(x[item])
-          )))
+          
+          lapply(op_detail$files,
```
[the remark this answers is not in the document]
Comment (ijlyttle) [reply R/boxr__internal_verb_exit.R:60]: I think this was the thing fouling up the tests (I suspect you came to the same conclusion).

My solution looks like:

```r
  file_list <- 
    purrr::map(
      items_list,
      function(item) {
        purrr::map_dfr(op_detail$files, ~.x[item])
      }
    )
```

But I have no idea if it is doing what Brandon intended...

[the remark this answers is not in the document]
Comment (nathancday) [reply R/boxr__internal_verb_exit.R:60]: I think this is essentially a function to stack the JSON elements from an API response into a data frame.

I found the error to be in `bind_rows()` (vectors related lower) when it wouldn't stack `octmode` and `character` values. I know `map_dfr()` calls `bind_rows()` so I guess we'd hit the same error.

The `as.character()` hack is yucky I agree. But since the Box API seems to take all of its inputs as strings. I figured the usability impact would be negligible.

Thanks for the review notes, I'll start responding tonight in bulk.

**Comment by nathancday** [comment]
I'm going to start moving these edits to a new PR, but will leave this one open until that is complete.

**Comment by ijlyttle** [comment]
I was able to use `usethis::pr_pull_upstream()` then sort through the conflicts. 

If you want to use this PR, great - if you want to move to a new PR, great.

File: R/boxr_misc.R [hunk 3]
```diff
@@ -276,3 +276,28 @@ boxDirCreate <- function(dir_name, parent_dir_id = box_getwd()) {
     terminate_on = box_terminal_http_codes()
   )
 }
+
+#'
+#' Open a browser to view a file or folder on Box's web interface
+#' 
+#' Thin wrapper of `browseURL` to make bouncing between R and Box a breeze.
+#' 
+#' @param file_id `character` a Box file ID
```
[the remark this answers is not in the document]
Comment (nathancday) [reply R/boxr_misc.R:285]: I think this should be `box_browse(dir_id = NULL, file_id = NULL)` and follow the pattern of `box_collab_create()`, thoughts?

File: R/boxr_add_description.R [hunk 4]
```diff
@@ -1,13 +1,19 @@
-#' Add description to a Box file
+#' Add description or comment to a Box file
```
[the remark this answers is not in the document]
Comment (nathancday) [reply R/boxr_add_description.R:1]: I think this should be `box_comment_create()` and `box_comment_get()` now following the `box_collab_X()` pattern

File: R/boxr_save_load.R [hunk 5]
```diff
@@ -59,6 +62,16 @@ box_save_image <- function(dir_id = box_getwd(), file_name = ".RData",
...
 }
 
+#' @rdname box_save
+#' @export
+box_save_rds <- function(object, dir_id = box_getwd(), file_name = ".RDS", ...,
+                     description = NULL) {
+  
+  temp_file <- normalizePath(file.path(tempdir(), file_name), mustWork = FALSE)
+  saveRDS(object, temp_file)
+  box_ul(as.integer(dir_id), temp_file, description = description)
```
[the remark this answers is not in the document]
Comment (nathancday) [reply R/boxr_save_load.R:72]: Good catch!

File: R/boxr_save_load.R [hunk 6]
```diff
@@ -59,6 +62,16 @@ box_save_image <- function(dir_id = box_getwd(), file_name = ".RData",
   box_ul(dir_id, temp_file, description = description)
 }
 
+#' @rdname box_save
+#' @export
+box_save_rds <- function(object, dir_id = box_getwd(), file_name = ".RDS", ...,
+                     description = NULL) {
+  
+  temp_file <- normalizePath(file.path(tempdir(), file_name), mustWork = FALSE)
```
[the remark this answers is not in the document]
Comment (nathancday) [reply R/boxr_save_load.R:70]: totally, using `fs::path_temp()` like we did to fix the test writing outside of temp_dir

File: R/boxr_file_versions.R (hunk 1 above)
Comment (nathancday) [reply R/boxr_file_versions.R:14]: I dropped this to `box_version(file_id)` in my responses. It feels succinct and fits with box_previous_versions()

**Comment by nathancday** [comment]
Hey @ijlyttle I think I responded to your edits on this PR.

There are a few that are still open for discussion and I tried to leave a specific comment above. All of these revolve around naming functions, our new favorite boxr past-time LOL

File: R/boxr_file_versions.R (hunk 1 above)
Comment (ijlyttle) [reply R/boxr_file_versions.R]: Sounds good! I wonder if this is a good opportunity to supersede, rather than to deprecate. 

Let me see if I can take a whack at it - here's an example of the idea: https://github.com/tidyverse/dplyr/blob/master/R/sample.R

File: R/box_comments.R [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    glue::glue("https://api.box.com/2.0/files/{file_id}/comments"),
+    get_token(),
+    terminate_on = box_terminal_http_codes()
+  )
+  
+  httr::stop_for_status(req)
+  
+  httr::content(req) %>% 
+    # should this be lifted in a helper, stack_list_element_into_dataframe()
+    .[["entries"]] %>%
```
Comment (ijlyttle) [inline R/box_comments.R:64]: The `.` is causing a NOTE. Could you use `purrr::pluck()` instead?

As well, perhaps could the file name be `boxr_comment.R`?

**Comment by ijlyttle** [comment]
I ended up following a much-easier path: `usethis::use_lifecycle()`.

I reworked the documentation to separate `box_previous_versions()`, then split apart the code in case we want to do something different with how `box_version_history()` behaves.

File: R/box_comments.R (hunk 7 above)
Comment (nathancday) [reply R/box_comments.R]: fixed

File: R/box_comment.R [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+#' \@user pattern is *not* supported at this time.
+#' @param comment_id `numeric` or `character`, comment ID at Box. 
+#' @return Invisible, the `comment_id` of the newly created Box comment.
+#' @examples 
+#' \dontrun{
+#' box_comment(12345, "Rmd report is ready for review")
+#' box_comment(comment_id = 98756, msg = "Responsed to your change requests!")
+#' }
+#' @export
+box_comment_create <- function(file_id = NULL, msg, comment_id = NULL) {
```
Comment (ijlyttle) [inline R/box_comment.R:20]: one point that I should have picked-up earlier, would you consider naming the argument `comment`, vs. `msg`?

Comment (nathancday) [reply R/box_comment.R:20]: I'm not set on `msg` but two things I considered: 
1) the api calls it "message"
2) down the road I would like to support another param `tag_users` that would insert the proper \@ tags as a prefix/suffix for notifying folks. This will require some accessory function to look people up by email-id, because box comments require the userID to be referenced. So I see the message text being combined with the tags to create a comment and wanted to have a clear distinction there. Currently inline \@-ing someone in the message seems like it is will be hard to pull off

Comment (ijlyttle) [reply R/box_comment.R:20]: I think what you propose could be managed with a single parameter (maybe `message`?) 

For now `msg` has to be `character`.

In the future we could do something like:

```r
msg = box_tag("Looks good!", user_id = 12345)

msg = box_tag("Looks good!", box_search_user("Nathan Day"))
```

In either case, `box_tag()` could add an S3 class (maybe called `boxr_tag`) so that `box_comment_create()` would know to use the `tagged_message` parameter.

There would need to be an internal helper that would call the Box user-endpoint such that:

```r
box_create_tag <- function(user_id){
  # call user endpoint to get name

  glue::glue("[{user_id}:{name}]")
}
```

**Comment by nathancday** [comment]
I'm not sure I like classing things anymore, after seeing how you'd have to use `box_comment_get()` with the class changes I just pushed to this branch.  It feels like we are making it clunkier for users by requiring an extra `as.frame()` call to get a useful object without offering any extra utility in the intermediate object.

Here I've layered our box_r class on top of the `list` from `httr::content()`, so the intermediate (pre-frame coerce) object can be name checked, subsetted etc. If we want to continue classing and as-ing everything I think we should make `boxr_collab` classes work like this too.

File: R/box_comment.R (hunk 8 above)
Comment (nathancday) [reply R/box_comment.R]: I've renamed the param to `message`

File: R/boxr_s3_classes.R [hunk 9]
```diff
@@ -356,6 +356,55 @@ print.boxr_collab_list <- function(x, ...) {
...
+  invisible(x)
+}
+
+
+# * Get -------------------------------------------------------------------
+
+#' @export
+#' 
+as.data.frame.boxr_comment_get <- function(x, ...) {
+  # stacking the comment records into a frame, one row per comment record
```
[the remark this answers is not in the document]
Comment (ijlyttle) [reply R/boxr_s3_classes.R:384]: As well, I think it would be good to have an `as_tibble()` method.

File: R/boxr_s3_classes.R [hunk 10]
```diff
@@ -356,6 +356,55 @@ print.boxr_collab_list <- function(x, ...) {
...
+  
+  # ability to reference parent fileID (from API call)
+  ret[["file_id"]] <- x$file_id
+  ret
+}
+
+#' @export
+#' 
+print.boxr_comment_get <- function(x, ...) {
+  x <- as.data.frame(x)
```
Comment (ijlyttle) [inline R/boxr_s3_classes.R:396]: I think `print_dispatch()` could work here, too.

**Review by ijlyttle — COMMENTED** [review]
To answer your question on the usefulness of returning an amorphous list-object of questionable utility, I think we should do so for two reasons:

- it fits in with how the other functions work.
- if someone wants a peek at what the API returned, perhaps to debug, it's the *only* way we can provide that. 

boxr was set up somewhat along the lines of what [httr suggests](https://httr.r-lib.org/articles/api-packages.html#return-a-helpful-object-1)

Clearly, this comes at a cost. But it also provides some flexibility in what someone can use - for a given box object (say a comment or list of comments), we can provide a way to extract:

- a data frame
- a tibble, if you like
- an opinionated optimal version of the return object, using something like `box_wrangle()` (in the future).

I know I have been burned by other boxr functions when I would print them, then expect a data frame. For this reason, I added the `--- printing as data frame ---` to the print method for collabs.

I see a lot of the pattern where Box either returns a single object (`box_create_comment()`) or a collection of objects in the `entries` element (`box_get_comment()`) - so we can provide a single-row data frame or a multi-row data frame. The pattern extends to collaborations and to versions, and I'm sure to other stuff.

This is where I thought a set of internal helpers could be reused: `stack_row_df|tbl()` and `stack_rows_df|tbl()`. By having these as internal functions, we can modify and test easily when new edge-cases appear.

The print helper `print_dispatch()` automatically tells you if {tibble} is loaded or not - if a person wanted to always print `as.data.frame`, we could set an option for that (I'll make a PR). 

I think there are better names out there for these internal functions, but being internal functions, we can rename them later.

What do you think?

File: tests/testthat/test_14_comments.R [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  fr1 <- box_write(data.frame("This"), "file1.txt")
+  
+  msg <- "hi there"
+  
+  expect_message(
+    resp <- box_comment_create(fr1$id, msg),
+    "Comment"
+  )
+  expect_s3_class(resp, "boxr_comment_create")
+  expect_s3_class(resp, "list")
```
Comment (ijlyttle) [inline tests/testthat/test_14_comments.R:45]: I forgot to mention this earlier: in the new testthat this will have to be `expect_type(resp, "list")`, which works now - I'm sure there will be other stuff that comes up when we switch over (not now!), but it could be useful to get ahead of this one.

**Comment by nathancday** [comment]
I agree about preserving the original API response as much as possible and I think that response of more useful in a data frame format (and most users will end up there eventually) `box_previous_versions()` has this data frame direct pattern. In a future I could see all `boxr` functions returning special frames, like `googledrive`'s `dribbles`, and if folks want to work with API response objects directly maybe they are the users of `cardboard`.

I just responded to changes and would really like to get this PR to done. Let me know if you are okay to merge and move on to other tasks.

**Comment by ijlyttle** [comment]
Sorry for delay - last week was busy at work.

I agree that `cardboard` is the way to go to support explicit access to the low-level API response, and that there should be a an easy way to get a high-level response without `as.data.frame()` or equivalent. 

I agree `box_previous_versions()` breaks that pattern, and we inherited that break. That said, I am uncomfortable to have a set of functions behave one way, another set behave a different way, and not have a good heuristic to tell which is which. 

Now that I think I have a better idea of the (admittedly less-than-ideal) compromise, I can make an effort to document it better, as it was not immediately evident to me. 

I should have some time this week to go through this and make any adjustments, so we can move on to other tasks.

**Comment by nathancday** [comment]
No worries, Box doesn't pay the bills. 

We can leave the behavior as is for `previous_versions` (hesitant to break long stable code) and support the list to coherce in `box_version_history`, does that sound like a fix?

FWIW I was imaging the return object reckoning (everyone gets a data.frame/tibble by default?) in v4.0.

**Comment by ijlyttle** [comment]
I agree we will have the object reckoning leading to v4 - I am in favor of the data.frame/tibble return (although we will have to sort out which one, and nesting, etc.)

For me the trick will be to migrate from current to v4 without breaking anything. testthat's edition approach seems interesting, but I think even the Tidyverse team are still figuring out if/how to extend that approach.

Getting to work now to map out the current situation, and to see how it might relate to some of our other recently-opened issues.

**Comment by ijlyttle** [comment]
This is a placeholder for the narrative that I will write Monday.

**Comment by ijlyttle** [comment]
Here's what I think I know about the changes:

- tweaked the documentation for `box_browse()`, `box_save_rds()`, box_read_rds()`.
- added `on.exit()` to clean up after temp files. I suspect there's a lot more of this to look at around the codebase; {withr} is coming out with some tools that could be handy.
- added {jsonlite} to Imports - I have no idea why that was not already there 😳
- tweaked the documentation for `box_comment_create()`, `box_comment_get()`, added to S3 documentation.
- `box_comment_get()` and `box_collab_get()` return only the content for the `entries` element of the API response. This is probably the biggest change I made:
  - the rest of the API response has only pagination information
  - to support pagination, we need to aggregate and return the `entries`
  - this is what we did with `box_ls()` (which was *my* first big PR to boxr, so forgetting this was 😳😳)
  - this also splashes on `stack_rows_df()` and `stack_rows_tbl()`
  - as a matter of semantics, distinguish between a `request`, `response`, the `content` of a response, and the `result` that we return.
- reworked the version functions; I think it is now looking a little bit how I might imagine a {cardboard} implementation (to be discussed, of course):
  - `box_version_history()` returns a data frame.
  - exported `box_version_api()` as an internal function; gives someone the chance to grab and parse the content themselves.
  - added the S3 argle-bargle.
  - made an internal parsing function `mutate_version_list()` to parse the entries. 
      - in the future, I can see some things like parsing datetimes, being part of a generalized parser used for all API content.
      - there are some things, like specifying the `version_no` column, that are particular to this class.
  - harmonized the `version_no` and `version_id` to agree with the forthcoming changes to `box_ls()`. 
  - also, fixed a bug in the version numbers; it was numbering the rows without adding `1` (that was a bug, right?)
- other documentation cleanup

I am happy with where this is now - if there's anything I should change or we should discuss, please let me know. Otherwise, I would be pleased that one of us "squash-and-merge".

 🙌
